packages/sdk/src/types/runtime/assets/assets-v2.md

```diff
@@ -0,0 +1,100 @@
+# Asset Unloading
+
+This proposal details the API changes required to allow developers to explicitly request
+that a given asset (material, texture, sound) be unloaded from memory on the clients,
+and unassign all references to them.
+
+Today, assets can be created individually, or created en masse by loading a container
+(glTF) for assets. This proposal does not have separate unload APIs for these two cases.
+Instead, during synchronization, the adapter will examine which assets are loaded at sync
+time, send a container load request if at least one asset from that container is still
+live, then send unloads for all the unused ones. This is processor-inefficient, but
+maintains API focus on the concept of assets.
+
+An alternative to the above would be to modify `AssetManager.loadGltf` to allow the
+developer to specify which assets are loaded. This is technically possible and solves
+the inefficiencies of the above, but requires developer knowledge of glTF internals,
+which may not be feasible. As another alternative, the API could require that
+container-loaded assets be unloaded as a block. This alternative doesn't really add
+any new functionality to the proposal, but reduces multipeer adapter complexity.
+
+## Proposed API Changes
+
+1. `AssetManager.unloadAssets(...assetIds: Asset[]): Promise<void>`
+
+	This method does four things:
+
+	1. It breaks all references to each asset by using the asset's internal reference
+		tracker, which will store references to each actor/asset that uses the asset.
+	2. It waits for the next batch of updates to propagate to the clients, so the assets
+		to be unloaded are not in use during the unload. This wait is done via a new
+		context method that resolves a promise when the next update is done.
+	3. It sends an `unload-assets` message to clients with the asset IDs. This will
+		trigger an `Object.Destroy` in Unity's asset cache, freeing the memory.
+	4. It collects success/failure messages from the client, and propagates any errors
+		from the client unload.
+
+2. `AssetManager.unloadGroup(group: AssetGroup): Promise<void>`
+
+	This simply calls `unloadAssets` with the IDs of all the assets in the group. If
+	alternative two is implemented, this will be the only way to unload container-loaded
+	assets. Even if it isn't, this will be more efficient than piecemeal unloading, but
+	the developer can decide which is more appropriate.
+
+3. `AssetManager.unloadUnused(): Promise<void>`
+
+	Loops over all assets, finds those that have no references, and calls `unloadAssets`
+	with their IDs.
+
+4. `Asset.unload(): Promise<void>`
```

Comment:
>: Promise<void> [](start = 18, length = 15)

Do we actually need to make these return Promise<void>? are there any cases you may want to wait on this?

Reply:
Hmm, probably not. The only use case that comes to mind is if you want to wait until after unload to load in new stuff.

---
In reply to: [300189356](https://github.com/microsoft/mixed-reality-extension-sdk/pull/369#discussion_r300189356) [](ancestors = 300189356)